File: 98. Validate Binary Search Tree.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def is_valis_bst(self, root: Optional[TreeNode]) -> bool:
        def check(root, low, high):
            if not root:
                return True

            if low and low.val >= root.val or high and high.val <= root.val:
                return False

            # left and right
            return check(root.left, low ,root) and check(root.right, root, high)
        return check(root, None, None)


if __name__ == "__main__":
    s = Solution()
    tree = TreeNode(2, TreeNode(1), TreeNode(3))
    print(s.is_valis_bst(tree))  # True
    tree = TreeNode(5, TreeNode(1), TreeNode(4, TreeNode(3), TreeNode(6)))
    print(s.is_valis_bst(tree))  # False
    tree = TreeNode(5, TreeNode(4), TreeNode(6, TreeNode(3), TreeNode(7)))
    print(s.is_valis_bst(tree))  # False
    tree = TreeNode(5, TreeNode(4), TreeNode(6, TreeNode(8, TreeNode(3), TreeNode(7))))
    print(s.is_valis_bst(tree))  # False

File: cpp/trees/98. Validate Binary Search Tree.cpp
#include <limits>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isValidBST(TreeNode* root, long minBound = LONG_MIN, long maxBound = LONG_MAX) {
        if (!root) return true;

        // Check if the current node's value is within valid bounds
        if (root->val <= minBound || root->val >= maxBound)
            return false;

        // Recursively validate left and right subtrees
        return isValidBST(root->left, minBound, root->val) &&
               isValidBST(root->right, root->val, maxBound);
    }
};